docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
+comprehensively represent HLO ABI where e.g. `T`, `tuple<T>` and
+`tuple<tuple<T>>` may be materially different depending on a particular
+implementation.
+
+```ebnf
+ElementType ::= BooleanType | IntegerType | FloatType | ComplexType
+BooleanType ::= 'i1'
+IntegerType ::= 'si4' | 'si8' | 'si16' | 'si32' | 'si64'
+              | 'ui4' | 'ui8' | 'ui16' | 'ui32' | 'ui64'
+FloatType   ::= 'f8E4M3FN' | 'f8E5M2' | 'bf16' | 'f16' | 'f32' | 'f64'
+ComplexType ::= 'complex' '<' ('f32' | 'f64') '>'
+```
 
-### Point-to-point communication
+**Element types** represent elements of tensor types. Unlike in many programming
+languages, these types are not first class in StableHLO. This means that
+StableHLO programs cannot directly represent values of these types (as a result,
+it is idiomatic to represent scalar values of type `T` with 0-dimensional tensor
+values of type `tensor<T>`).
+
+  * **Boolean type** represents boolean values `true` and `false`.
+  * **Integer types** can be either signed (`si`) or unsigned (`ui`) and have
+    one of the supported bit widths (`4`, `8`, `16`, `32` or `64`).
+    Signed `siN` types represent integer values from `-2^(N-1)` to `2^(N-1)-1`
+    inclusive, and unsigned `uiN` types represent integer values from `0` to
+    `2^N-1` inclusive.
+  * **Floating-point types** can be one of the following:
+    * `f8E4M3FN` and `f8E5M2` types corresponding to respectively the
+    `E4M3` and `E5M2` encodings of the FP8 format described in
+    [FP8 Formats for Deep Learning](https://arxiv.org/abs/2209.05433).
+    * `bf16` type corresponding to the `bfloat16` format described in
+    [BFloat16: The secret to high performance on Cloud TPUs](https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus).
+    * `f16`, `f32` and `f64` types corresponding to respectively
+    `binary16` ("half precision"), `binary32` ("single precision") and
+    `binary64` ("double precision") formats described in
+    [the IEEE 754 standard](https://ieeexplore.ieee.org/document/8766229).
+  * **Complex types** represent complex values that have a **real part**
+    and an **imaginary part** of the same **element type**. Supported complex
+    types are `complex<f32>` (both parts are of type `f32`) and `complex<f64>`
+    (both parts are of type `f64`).
+
+```ebnf
+FunctionType ::= '(' [ValueType {',' ValueType}] ')' '->' '(' [ValueType {',' ValueType}] ')'
+```
 
-StableHLO processes can communicate with each other through
-**StableHLO channels**. A channel is represented by a positive id of type
-`si64`. Through various ops, it is possible to send values to channels and
-receive them from channels.
+**Function types** represent both named and anonymous functions. They have
+input types (the list of types on the left-hand side of `->`) and output types
+(the list of types on the right-hand side of `->`). In many programming
+languages, function types are first class, but not in StableHLO.
```

Comment:
Via reflection / metaprogramming API's? This represents the functions signature IIUC